pallets/xcm-rate-limiter/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	/// Accumulated amounts for each asset
+	#[pallet::getter(fn accumulated_amount)]
+	pub type AccumulatedAmounts<T: Config> =
+		StorageMap<_, Blake2_128Concat, MultiLocation, AccumulatedAmount, ValueQuery>;
+
+	#[pallet::event]
+	pub enum Event<T: Config> {}
+
+	#[pallet::error]
+	#[cfg_attr(test, derive(PartialEq, Eq))]
+	pub enum Error<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+}
+
+fn get_loc_and_amount(m: &MultiAsset) -> Option<(MultiLocation, u128)> {
+	match m.id {
+		AssetId::Concrete(location) => match m.fun {
+			Fungibility::Fungible(amount) => Some((location, amount)),
+			_ => None,
+		},
+		_ => None,
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn get_locations_and_amounts(instruction: &Instruction<T::RuntimeCall>) -> Vec<(MultiLocation, u128)> {
+		use Instruction::*;
+		match instruction {
+			// NOTE: This does not address the native asset "coming back" from other chains.
+			ReserveAssetDeposited(multi_assets) | ReceiveTeleportedAsset(multi_assets) => {
+				multi_assets.inner().iter().flat_map(get_loc_and_amount).collect()
+			}
+			_ => Vec::new(),
+		}
+	}
+}
+
+impl<T: Config> XcmDeferFilter<T::RuntimeCall> for Pallet<T> {
+	fn deferred_by(
+		_para: polkadot_parachain::primitives::Id,
+		_sent_at: RelayChainBlockNumber,
+		versioned_xcm: &VersionedXcm<T::RuntimeCall>,
+	) -> Option<RelayChainBlockNumber> {
+		use xcm::IntoVersion;
+		let maybe_xcm = versioned_xcm.clone().into_version(3);
+		let Ok(V3(xcm)) = maybe_xcm else { return Some(T::MaxDeferDuration::get()) };
+		// SAFETY NOTE: It is fine to only look at the first instruction because that is how assets will arrive on chain.
+		//              This is guaranteed by `AllowTopLevelExecution` which is standard in the ecosystem.
+		let Some(instruction) = xcm.first() else { return None };
+		for (location, amount) in Pallet::<T>::get_locations_and_amounts(instruction) {
+			let accumulated_liquidity = AccumulatedAmounts::<T>::get(location);
+
+			let Some(asset_id) = T::CurrencyIdConvert::convert(location) else { continue };
+			let Some(limit_per_duration) = T::RateLimitFor::get(&asset_id) else { continue };
+			let defer_duration = T::DeferDuration::get();
+
+			let current_time = T::RelayBlockNumberProvider::current_block_number();
+			let time_difference = current_time.saturating_sub(accumulated_liquidity.last_updated);
+
+			let new_accumulated_amount = calculate_new_accumulated_amount(
+				defer_duration.saturated_into(),
+				limit_per_duration,
+				amount,
+				accumulated_liquidity.amount,
+				time_difference.saturated_into(),
+			);
+
+			let deferred_by = calculate_deferred_duration(
+				defer_duration.saturated_into(),
+				limit_per_duration,
+				new_accumulated_amount,
+			);
+
+			AccumulatedAmounts::<T>::insert(
+				location,
+				AccumulatedAmount {
+					amount: new_accumulated_amount,
+					last_updated: current_time,
+				},
+			);
+
+			if deferred_by > 0 {
```

Comment:
small thing, but here we could use a guard statement for returning None, then returning Some in the next line.

By doing so we can get rid of the `else` stament.